node/src/components/consensus/protocols/highway.rs

```diff
@@ -153,7 +153,11 @@ impl<I: NodeIdT, C: Context + 'static> HighwayProtocol<I, C> {
             }
             AvEffect::ScheduleTimer(timestamp) => vec![ProtocolOutcome::ScheduleTimer(timestamp)],
             AvEffect::RequestNewBlock(block_context) => {
-                vec![ProtocolOutcome::CreateNewBlock { block_context }]
+                let past_values = self.non_finalized_values().cloned().collect();
```

Comment:
It would be even safer to put the fork choice itself into `AvEffect::RequestNewBlock`. Then even if the same `State` method that created that effect made any changes afterwards, we'd still get the right values. (Not sure if that can happen currently.)

Reply:
And then I'd pass that fork choice into `non_finalized_values`, so that the outcome doesn't depend on _when_ you call `non_finalized_values` anymore.